RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforGPU.h

```diff
@@ -0,0 +1,342 @@
+#ifndef RecoLocalTracker_SiPixelRecHits_pixelCPEforGPU_h
+#define RecoLocalTracker_SiPixelRecHits_pixelCPEforGPU_h
+
+#include <cassert>
+#include <cmath>
+#include <cstdint>
+#include <iterator>
+
+#include "CUDADataFormats/SiPixelCluster/interface/gpuClusteringConstants.h"
+#include "DataFormats/GeometrySurface/interface/SOARotation.h"
+#include "Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/cudaCompat.h"
+
+namespace pixelCPEforGPU {
+
+  using Frame = SOAFrame<float>;
+  using Rotation = SOARotation<float>;
+
+  // all modules are identical!
+  struct CommonParams {
+    float theThicknessB;
+    float theThicknessE;
+    float thePitchX;
+    float thePitchY;
+  };
+
+  struct DetParams {
+    bool isBarrel;
+    bool isPosZ;
+    uint16_t layer;
+    uint16_t index;
+    uint32_t rawId;
+
+    float shiftX;
+    float shiftY;
+    float chargeWidthX;
+    float chargeWidthY;
+    uint16_t pixmx;  // max pix charge
+
+    float x0, y0, z0;  // the vertex in the local coord of the detector
+
+    float sx[3], sy[3];  // the errors...
+
+    Frame frame;
+  };
+
+  using phase1PixelTopology::AverageGeometry;
+
+  struct LayerGeometry {
+    uint32_t layerStart[phase1PixelTopology::numberOfLayers + 1];
+    uint8_t layer[phase1PixelTopology::layerIndexSize];
+  };
+
+  struct ParamsOnGPU {
+    CommonParams const* m_commonParams;
```

Comment:
`commonParams()` *etc.* are already used for the `__restrict__` accessors below.

Making the members `private` would break (at least) the memory handling in RecoLocalTracker/SiPixelRecHits/src/PixelCPEFast.cc .

I would suggest leaving this for a follow up PR.

Reply:
Added to https://github.com/cms-sw/cmssw/issues/32483 .